### ISSUE TITLE
Use bootupd and remove ostree-grub2 (was: All deployments are shown twice in grub)

### ISSUE THREAD

**Issue by YaLTeR** [fedora-silverblue/issue-tracker#120]
# Updated issue

Bootupd is now included and enabled by default on Atomic Desktops: https://fedoraproject.org/wiki/Changes/FedoraSilverblueBootupd

The duplicate entry part of this issue remains and will be fixed once we move to a static GRUB config. This should happen alongside the composefs migration. See: https://gitlab.com/fedora/ostree/sig/-/issues/35

---

# Original issue

![boot-entry-issue](https://user-images.githubusercontent.com/1794388/103624021-febdf180-4f49-11eb-9617-a1e6e049277a.jpeg)

F33, ThinkPad T495s

**Comment by felipeborges** [comment]
I can reproduce this too.

**Comment by jlebon** [comment]
See https://fedoraproject.org/wiki/Common_F31_bugs#On_Fedora_Silverblue.2FIoT.2C_the_GRUB_menu_shows_duplicate_entries (I've added an "Update:" section there).

**Comment by YaLTeR** [comment]
> See https://fedoraproject.org/wiki/Common_F31_bugs#On_Fedora_Silverblue.2FIoT.2C_the_GRUB_menu_shows_duplicate_entries (I've added an "Update:" section there).

So... should I migrate to BLS? If I should then maybe it should rather be done by default in Silverblue?

**Comment by jlebon** [comment]
> So... should I migrate to BLS? If I should then maybe it should rather be done by default in Silverblue?

If you can, yes. I'm not sure if it's worth forcing a migration at this point because of the potential risks (see e.g. https://discussion.fedoraproject.org/t/boot-entries-gone-after-upgrade/8026). And as mentioned in the wiki, the migration script only works on EFI because it might not be safe to do it automatically on BIOS (see https://github.com/ostreedev/ostree/pull/2044#issuecomment-608316436).

**Comment by Jmennius** [comment]
What about systems that did not upgrade to F31 (clean install F31+)?
I did not find any place that creates `/boot/grub2/.grub2-blscfg-supported` for new installations.
I have BLS enabled (Silverblue) but there is no such file (probably this system started with F31).

I am also concerned that `grub2-switch-to-blscfg` does not work well with https://fedoraproject.org/wiki/Changes/UnifyGrubConfig.

cc @jlebon @martinezjavier

**Comment by dustymabe** [comment]
I'm seeing this on a fresh install of F35 Silverblue. I guess we need to track down what used to create the `/boot/grub2/.grub2-blscfg-supported` and see why it's not getting created any longer.

**Comment by dustymabe** [comment]
Found it: https://src.fedoraproject.org/rpms/grub2/blob/9fdaa794e06846f1426bd15360401a8b1a5017a9/f/0065-Add-grub2-switch-to-blscfg.patch#_362

**Comment by jlebon** [comment]
> I'm seeing this on a fresh install of F35 Silverblue. I guess we need to track down what used to create the `/boot/grub2/.grub2-blscfg-supported` and see why it's not getting created any longer.

I think... nothing ever actually created that for new installs. The idea we discussed was to have Anaconda do it, but it looks like we never actually followed through with it.

The other option is to just assume GRUB is new enough, but that can break people upgrading from an old install with old GRUB. The failure mode here is particularly bad: they'll get no boot entries.

Maybe we just need to be really loud about it and say that you need to make sure that your GRUB is up to date before $date using either the migration script or bootupd, and then we rip out this check.

**Comment by jlebon** [comment]
> Maybe we just need to be really loud about it and say that you need to make sure that your GRUB is up to date before $date using either the migration script or bootupd, and then we rip out this check.

A prerequisite for that is of course to ship bootupd in FSB, which we currently don't.

**Comment by jlebon** [comment]
Yeah, I think bootupd is probably the cleanest path forward for this and for possible future issues like it. So my strawman is something like:
- add bootupd in FSB
- teach Anaconda to call out to `bootupctl`
- teach `15_ostree` to read bootupd metadata to know if GRUB supports `blscfg` (but really, if bootupd metadata exists, it surely does, so this might just be `test -f /boot/.bootupd-state.json`)
- tell upgrading users that this bug is fixed in the latest bootloader, and they can update using `bootupctl adopt-and-update`

@cgwalters @martinezjavier WDYT?

**Comment by travier** [comment]
This might need a Fedora Change for F36 (Silverblue & Kinoite).

**Comment by martinezjavier** [comment]
> @cgwalters @martinezjavier WDYT?

This makes sense to me. In fact, it would be great if even the non-ostree Fedora variants could adopt `bootupd`.

**Comment by travier** [comment]
The change deadline is mid January if we think this is a self contained change: https://fedorapeople.org/groups/schedule/f-36/f-36-key-tasks.html

**Comment by ormandj** [comment]
This would be quite a nice change - I was surprised to see the duplicate entries on a clean 35 Silverblue installation, and ran across this issue while researching resolution.

**Comment by travier** [comment]
How do all of you feel about trying to make that happen for F37 and maybe opt-in for F36 (ship bootupd but not enabled by default)? I can start the change request.

**Comment by tpopela** [comment]
It was great to see @martinezjavier 's option and if he's in favor, then I'm as well. Thank you @travier for looking into this!

**Comment by jlebon** [comment]
> How do all of you feel about trying to make that happen for F37 and maybe opt-in for F36 (ship bootupd but not enabled by default)? I can start the change request.

SGTM, thanks for picking this up!

**Comment by tpopela** [comment]
It would be good to tackle this after the recent round of the grub2 CVEs. We might also need to fix https://github.com/coreos/bootupd/issues/108 . But it looks like there was some work done by @cgwalters , but might be not finished - what's the current state of it Colin?

**Comment by koko-ng** [comment]
I ran into this problem while debugging why firmware updates weren't applied. If anyone runs into the same issue, before there is a cleaner fix it's possible to update shim like this:

```
sudo rpm-ostree usroverlay 
wget https://kojipkgs.fedoraproject.org//packages/shim/15.6/1/x86_64/shim-x64-15.6-1.x86_64.rpm
sudo rpm -i  --reinstall shim-x64-15.6-1.x86_64.rpm
```

**Comment by dustymabe** [comment]
I'm not sure if `rpm-ostree usroverlay` is what you want:

```
       usroverlay
           Mount a writable overlay filesystem on /usr which is
           active only for the remainder of the system boot. This
           is intended for development, testing, and debugging.
           Changes will not persist across upgrades, or rebooting
           in general.
```

**Comment by koko-ng** [comment]
It's just so that rpm can write to the database, of course it will be discarded on next reboot, but the changes to /bin/efi will effectively be kept as they are not managed by ostree and as rpm reinstall the same version of the package, the content of the persistent database (the one managed by rpm-ostree) is now coherent with the content of /bin/efi.

**Comment by travier** [comment]
With https://pagure.io/workstation-ostree-config/pull-request/288, we now have bootupd in Rawhide but not enabled by default.

Let's track what's missing here and make a Change request for F38.

For reference, the commit message from Jonathan:
```
We need to make it easier to update the bootloader on these variants
because unlike on traditional systems, it's not updated automatically
with the rest of the system. Add bootupd for that.

This would allow fixing issues like:
- https://github.com/coreos/rpm-ostree/issues/3715
- https://github.com/fedora-silverblue/issue-tracker/issues/120#issuecomment-979268679

It won't be enabled by default and as mentioned in that comment requires
work in Anaconda to be seamless. But at least with this users should be
able to adopt and update:

https://github.com/coreos/bootupd/blob/main/README-design.md

See also the tracker issue where we did this for Fedora CoreOS:

https://github.com/coreos/fedora-coreos-tracker/issues/510
```

**Comment by travier** [comment]
With:
- https://pagure.io/workstation-ostree-config/pull-request/306
- https://pagure.io/workstation-ostree-config/pull-request/303
we now have bootupd in F36 & F37; still not enabled by default, but that should let folks try it and report feedback so that we can fix issues.

**Comment by jvnknvlgl** [comment]
This might be relevant here: https://github.com/coreos/bootupd/issues/378

I'm having the same issue, so I'm assuming bootupd cannot be used in Silverblue yet.

**Comment by travier** [comment]
Yes, we need to wait until this is fixed (and the package is updated).

**Comment by travier** [comment]
We still need to figure out the migration process to be able to remove ostree-grub2

**Comment by travier** [comment]
Automated migration would require: https://github.com/rhboot/shim/pull/502

**Comment by jmaibaum** [comment]
I have this issue trying to update the Secure boot dbx 77 -> 217 on Silverblue 37:

```
Blocked executable in the ESP, ensure grub and shim are up to date: /boot/efi/EFI/fedora/shimx64-fedora.efi Authenticode checksum [0ce02100f67c7ef85f4eed368f02bf7092380a3c23ca91fd7f19430d94b00c19] is present in dbx``

I tried the workaround from https://github.com/fedora-silverblue/issue-tracker/issues/120#issuecomment-1177570662:

> ```
> sudo rpm-ostree usroverlay 
> wget https://kojipkgs.fedoraproject.org//packages/shim/15.6/1/x86_64/shim-x64-15.6-1.x86_64.rpm
> sudo rpm -i  --reinstall shim-x64-15.6-1.x86_64.rpm
> ```

Using `shim-x64-15.6-2.x86_64.rpm` instead of `shim-x64-15.6-1.x86_64.rpm`, as this was what `rpm-ostree db list` repored for my install. `fwupdmgr upgrade` still fails with the same error, but now I also get a secure boot signature violation, so I had to disable secure boot for now. Is this even a valid workaround for the Secure Boot dbx update failing?

https://bugzilla.redhat.com/show_bug.cgi?id=2127995 seems to suggest that there is a workaround, which sounds similar to what I tried above, but so far it doesn't work.

**Comment by jmaibaum** [comment]
Moving `/boot/efi/EFI/fedora/shimx64-fedora.efi` helped getting the Secure Boot dbx update install successfully, yet I still can't reenable Secure Boot.

I am now trying to get `bootupd` to work:

```console
$ bootupctl status
error: connecting to /run/bootupd.sock: ENOENT: No such file or directory
```

**Comment by jmaibaum** [comment]
Got past the previous error by `systemctl enable bootupd.socket`, now I am at

```console
$ sudo bootupctl status
error: internal error: Could not find "/dev/disk/by-partlabel/EFI-SYSTEM"
```

**Comment by travier** [comment]
Working on the change for this issue in https://fedoraproject.org/wiki/Changes/FedoraSilverblueBootupd

**Comment by 0rzech** [comment]
@travier Is there any proper manual workaround for this we could apply before the change proposal you linked is implemented? Something that wouldn't require us to download files from the Internet.

**Comment by jmaibaum** [comment]
After rebasing to Fedora Silverblue 37, `bootupd` is now on v0.2.8, but I still can't continue:

```console
$ sudo bootupctl status
No components installed.
Detected: EFI: unknown
Boot method: EFI
$ sudo bootupctl adopt-and-update
error: internal error: Component EFI has no available update
```

**Comment by travier** [comment]
I don't have a workaround that I've tested. Help welcomed. Maybe https://github.com/fedora-silverblue/issue-tracker/issues/120#issuecomment-1177515110 can be updated for F37 and might work but I have not tested this.

**Comment by jmaibaum** [comment]
I just retried https://github.com/fedora-silverblue/issue-tracker/issues/120#issuecomment-1177515110 like this:

```
sudo rpm-ostree usroverlay 
wget https://kojipkgs.fedoraproject.org//packages/shim/15.6/2/x86_64/shim-x64-15.6-2.x86_64.rpm
sudo rpm -i  --reinstall shim-x64-15.6-2.x86_64.rpm
```

Followed by a reboot, and then I tried to let `bootupd` take over again, but it's still not detecting anything in the EFI:

```
$ sudo bootupctl status
No components installed.
Detected: EFI: unknown
Boot method: EFI
$ sudo bootupctl adopt-and-update
error: internal error: Component EFI has no available update
```

Same as before.

**Comment by cgwalters** [comment]
Hi, using bootupd requires it to be enabled at build time, see https://pagure.io/workstation-ostree-config/pull-request/308

**Comment by travier** [comment]
Adding some more references here for cross-linking:
* https://pagure.io/workstation-ostree-config/pull-request/156
* https://github.com/ostreedev/ostree/pull/1929

**Comment by travier** [comment]
This has been accepted: https://pagure.io/fesco/issue/2900
Pending approval for rpm-ostree unified core change now:  https://pagure.io/fesco/issue/2901

**Comment by travier** [comment]
Pushed to F39

**Comment by firefoxlover** [comment]
A current workaround for this is:

```
if [ -d /sys/firmware/efi ]; then
  grub2-switch-to-blscfg
  grub2-mkconfig -o /boot/efi/EFI/fedora/grub.cfg
else
  block_device=$(lsblk -spnlo name $(grub2-probe --target=device /boot/grub2) | tail -n1)
  grub2-install $block_device
  touch /boot/grub2/.grub2-blscfg-supported
  grub2-mkconfig -o /boot/grub2/grub.cfg
fi
```

**Comment by juhp** [comment]
> Pushed to F39

and F38?

**Comment by firefoxlover** [comment]
@juhp yes seems like a rather big problem with a small solution.

**Comment by travier** [comment]
https://github.com/coreos/fedora-coreos-tracker/issues/1441

**Comment by travier** [comment]
To clarify https://github.com/fedora-silverblue/issue-tracker/issues/120#issuecomment-1441638617: This has not landed yet because it's blocked by https://github.com/fedora-silverblue/issue-tracker/issues/333 which we had to report to F39.

**Comment by travier** [comment]
> A current workaround for this is:
> 
> ```
> if [ -d /sys/firmware/efi ]; then
>   grub2-switch-to-blscfg
>   grub2-mkconfig -o /boot/efi/EFI/fedora/grub.cfg
> else
>   block_device=$(lsblk -spnlo name $(grub2-probe --target=device /boot/grub2) | tail -n1)
>   grub2-install $block_device
>   touch /boot/grub2/.grub2-blscfg-supported
>   grub2-mkconfig -o /boot/grub2/grub.cfg
> fi
> ```

This will only works if your bootloader is recent enough to have BLS support. Otherwise you'll get no boot entries.

**Comment by juhp** [comment]
Sorry to persist: so does "pushed to f39" mean deferred? Or do you still expect this to land in F38?

(If deferred, the Change pages should be updated to reflect that)

**Comment by miabbott** [comment]
> Sorry to persist: so does "pushed to f39" mean deferred? Or do you still expect this to land in F38?
> 
> (If deferred, the Change pages should be updated to reflect that)

Deferred to F39.

**Comment by firefoxlover** [comment]
> > A current workaround for this is:
> > ```
> > if [ -d /sys/firmware/efi ]; then
> >   grub2-switch-to-blscfg
> >   grub2-mkconfig -o /boot/efi/EFI/fedora/grub.cfg
> > else
> >   block_device=$(lsblk -spnlo name $(grub2-probe --target=device /boot/grub2) | tail -n1)
> >   grub2-install $block_device
> >   touch /boot/grub2/.grub2-blscfg-supported
> >   grub2-mkconfig -o /boot/grub2/grub.cfg
> > fi
> > ```
> 
> This will only works if your bootloader is recent enough to have BLS support. Otherwise you'll get no boot entries.

Bootloader means grub right? Because I am on ooold Bios and it works perfectly. Grub should be up to date on Fedora or not?

**Comment by firefoxlover** [comment]
@travier Thanks for the clarification! Sorry when I dont really know the technical details, is there a manual way to update the bootloader?

**Comment by travier** [comment]
No manual way supported right now. There are workaround in the issue here and linked one. Everything is at your own risk.

**Comment by yann-soubeyrand** [comment]
Hello,

Inspired by https://github.com/fedora-silverblue/issue-tracker/issues/120#issuecomment-1177515110, I also did
```
sudo rpm-ostree usroverlay
wget https://kojipkgs.fedoraproject.org//packages/grub2/2.06/95.fc38/x86_64/grub2-efi-x64-2.06-95.fc38.x86_64.rpm
sudo rpm -i  --reinstall grub2-efi-x64-2.06-95.fc38.x86_64.rpm
```
and I was able to re-enable secure boot.

**Comment by jmaibaum** [comment]
Thanks @yann-soubeyrand , this did work for me now as well.

**Comment by travier** [comment]
I've merged in Rawhide: https://pagure.io/workstation-ostree-config/pull-request/330

My test images (https://gitlab.com/fedora/ostree/ci-test) also include this already.

**Comment by Verhoeckx** [comment]
> I've merged in Rawhide: https://pagure.io/workstation-ostree-config/pull-request/330
> 
> My test images (https://gitlab.com/fedora/ostree/ci-test) also include this already.

Will this be available in Fedora 39?

The solution[ in the above comment](https://github.com/fedora-silverblue/issue-tracker/issues/120#issuecomment-1519070229), didn't work for me. Will your merge request solve the problem?

**Comment by Verhoeckx** [comment]
@yann-soubeyrand 

I executed the three lines that you mention [in the above comment](https://github.com/fedora-silverblue/issue-tracker/issues/120#issuecomment-1519070229) and when I now execute (after a reboot) `rpm-ostree upgrade`, I get the following error: `error: loading sysroot: opendirat: Permission denied`. Do I need the execute some other command after the three lines you mention (to close the overlay)?

**Comment by juhp** [comment]
Is this still in current Silverblue/Kinoite?  I am not seeing it in either 39 or Rawhide, but if so what to look for?

**Comment by travier** [comment]
This is still planned to be included in F39 but not there yet due to https://pagure.io/workstation-ostree-config/pull-request/330#comment-190406 (the PR has been reverted).

**Comment by travier** [comment]
Pushed back to Fedora 40. Might come as an update to F39 if we backport the support.

**Comment by travier** [comment]
bootupd support in Anaconda got merged 🎉 
- https://github.com/rhinstaller/anaconda/pull/5298 -> https://github.com/rhinstaller/anaconda/pull/5342

**Comment by 0rzech** [comment]
> I executed the three lines that you mention [in the above comment](https://github.com/fedora-silverblue/issue-tracker/issues/120#issuecomment-1519070229) and when I now execute (after a reboot) `rpm-ostree upgrade`, I get the following error: `error: loading sysroot: opendirat: Permission denied`. Do I need the execute some other command after the three lines you mention (to close the overlay)?

@Verhoeckx Did you find any solotion to that? I've just stumbled upon this error too.

**Comment by travier** [comment]
See https://github.com/fedora-silverblue/issue-tracker/issues/543#issuecomment-2048350047 for a lightly tested set of commands to update your EFI bootloader until we have full bootupd support in Fedora Atomic Desktops.

**Comment by scottwillmoore** [comment]
I'm confused :confused:, did `bootupd` make it into 40, or was it deferred until 41? At least from the threads below it looks like it was pushed back to 41, but then was ready for 40, however I don't see `bootupd` on my installation? 

- https://fedoraproject.org/wiki/Changes/FedoraSilverblueBootupd
- https://bugzilla.redhat.com/show_bug.cgi?id=2150982

Under the assumption `bootupd` is not included with 40, is it possible to still use it?

**Comment by davidstrauss** [comment]
I researched that exact question about deferral earlier today. I found at least one comment (I think on Bugzilla) that confirmed deferral to F41. I also noticed that most F40 docs fail to reflect that.

**Comment by Malix-off** [comment]
Hello @travier, any update about the backport to F40?

**Comment by travier** [comment]
We can only *consider* backporting to F40 after it lands in F41. Status is tracked in https://gitlab.com/fedora/ostree/sig/-/issues/1

**Comment by travier** [comment]
With bootupd landing in Fedora 41, the first part of this issue is done. Now we have to wait for F42 to be able to remove the legacy ostree-grub2 package to complete this.

**Comment by davidstrauss** [comment]
Nice work, folks. I just updated to Fedora Silverblue 41, and it looks like `bootupd` is properly active:

```
straussd@phoenix:~$ sudo bootupctl adopt-and-update
Running as unit: bootupd.service
No components are adoptable.
straussd@phoenix:~$ sudo bootupctl status
Running as unit: bootupd.service
Component EFI
  Installed: grub2-efi-ia32-1:2.12-10.fc41.x86_64,grub2-efi-x64-1:2.12-10.fc41.x86_64,shim-ia32-15.8-3.x86_64,shim-x64-15.8-3.x86_64
  Update: At latest version
No components are adoptable.
Boot method: EFI
```

**Comment by travier** [comment]
Closed by mistake. This should be fixed in Rawhide, coming soon to F42.

**Comment by travier** [comment]
I've pushed this to Fedora 42.
